chore(performance): Only set state if differences were found

k88hudson opened the pull request:
This PR changes the page worker behaviour so that it only saves a new copy of the redux state if anything has changed.

k88hudson commented:
@dmose tests are experiencing intermittent failures at the moment

dmose commented:
Travis seems to have failed because an intermittent that @ncloudioj has already picked up.  I restarted the travis build, so assuming it succeeds, you're all set...

dmose commented:
The only nit referred to by the label is CI tests passing.

k88hudson commented:
@dmose  I changed the strategy a little here, moving to strict equality to save on performance (`redux-watch` basically does `===` on each reducer). This also required a slight change in one of our reducers, which was accidentally creating new objects. I added a test to prevent that from happening in the future.

dmose commented on common/reducers/Filter.js:
```diff
@@ -11,7 +11,8 @@ module.exports = function Filter(prevState = INITIAL_STATE, action) {
   switch (action.type) {
```
It seems like we could do even better here by only creating the state object inside the NOTIFY_FILTER_QUERY, so that it never gets created at all in the default cause, since it will be unused...

dmose replied:
I belatedly remembered that you can't put a declaration inside a case block.  That said, you could declare it using let above the switch, and Object.assign inside the case block.  Totally a micro-optimization though, so it's entirely up to you whether you do it.  If you do so, I'd suggest adding a comment explaining why so that some future maintainer doesn't undo it.

dmose commented on addon/PageWorker.js:
```diff
@@ -27,7 +28,8 @@ class PageWorker {
     // returns a function which will unsubscribe
-    this._unsubscribe = this._store.subscribe(this._onDispatch);
+    const w = watch(this._store.getState);
+    this._unsubscribe = this._store.subscribe(w(this._onDispatch));
```
I'm confused.  Since _onDispatch doesn't use any of the arguments that redux-watch has w pass, what is the point of adding redux-watch?

k88hudson replied:
watch does an equality check that doesn't call the callback if `prevState === newState`

k88hudson replied:
I will add a comment

dmose replied:
Great; thanks

dmose commented on content-test/addon/PageWorker.test.js:
```diff
@@ -39,19 +39,23 @@ describe("PageWorker", () => {
-      assert(store.getState(), 3);
       pageWorker.connect();
-      pageWorker._onDispatch();
+      store.dispatch({type: "add", number: 3});
```
Nice job removing the impl detail from the test!

k88hudson commented:
@dmose is this good to merge?

dmose commented on content-test/common/reducers/equality.test.js:
```diff
@@ -0,0 +1,35 @@
+    store = createStore({logger: false});
+  });
+  it("should return a reference", () => {
+    assert.equal(store.getState(), store.getState());
```
Should this be assert.strictEqual?

k88hudson replied:
in chai `equal` is strict equality as opposed to `deepEqual`

dmose replied:
If I'm reading the reference at http://chaijs.com/api/assert/ correctly, .equal is non-strict equality, and .strictEqual is what we want here...

dmose commented on content-test/common/reducers/equality.test.js:
```diff
@@ -0,0 +1,35 @@
+  it("should return a reference", () => {
+    assert.equal(store.getState(), store.getState());
+  });
+  it("should dispatch rows", () => {
```
The title here suggests that the test wants to test the dispatchRows function above, which it obviously doesn't.  

It seems like this is really testing that TOP_FRECENT_SITES_RESPONSE maps to newState.TopSites.rows, but perhaps I'm misinterpreting.

I think something wants to change here; it _might_ just be the description of the test.

dmose commented on content-test/common/reducers/equality.test.js:
```diff
@@ -0,0 +1,35 @@
+const {actions} = require("common/action-manager");
+const {createRows} = require("test/faker");
+
+describe("Reducer equality", () => {
```
There's something incongruent about this title.  All of the "it" clauses are not specifically referring to "Reducer equality", I don't think.  Perhaps this doesn't matter; perhaps there's another phrasing that would improve readability.

k88hudson commented:
@dmose ok, added some review fixes and filed https://github.com/mozilla/activity-stream/issues/1463 and https://github.com/mozilla/activity-stream/issues/1462

dmose commented:
Looks good; thanks for the tweaks!  r=@dmose
